pmlc/dialect/stripe/autostencil_pass.cc

```diff
@@ -0,0 +1,312 @@
+// Copyright 2019, Intel Corporation
+
+#include "pmlc/dialect/stripe/transcode.h"
+#include "pmlc/dialect/stripe/autostencil_pass.h"
+
+#include <sstream>
+#include <vector>
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/Translation.h"
+
+#include "pmlc/dialect/stripe/analysis.h"
+#include "pmlc/dialect/stripe/dialect.h"
+#include "pmlc/dialect/stripe/ops.h"
+#include "pmlc/dialect/stripe/transforms.h"
+#include "pmlc/dialect/stripe/util.h"
+
+#include "base/util/logging.h"
+
+namespace pmlc {
+namespace dialect {
+namespace stripe {
+
+// Number of tensors for the matrix multiplication
+const size_t numTensors = 3;
+// Number of searching index, i.e., M, N, K
+const size_t numIndex = 3;
+// Maximum number of index in a block, e.g., ParallelForOp
+const size_t indexLimit = 32;
+
+typedef llvm::SmallPtrSet<BlockArgument*, indexLimit> BlockArgumentSet;
+
+class AutoStencil {
+public:
+  explicit AutoStencil(size_t processors);
+  // Main function
+  void Stencil(ParallelForOp op);
+private:
+  // The throughput of M*N*K matrix multiplication
+  double Throughput(size_t m, size_t n, size_t k);
+  // Evaluate the performance of the current searching state
+  double Evaluate();
+  // Search the index, i.e., M, N, K, for the inner block
+  void SearchIndex(size_t matrix_idx);
+  // For (M, N, K) in the inner block, search their tiles
+  void SearchTiles(size_t idx);
+  // Transform the current ParallelForOp
+  void Transform();
+
+  // Collect the index used by value's RefineOp
+  BlockArgumentSet RefUsedIdxs(Value* value, bool with_conflict);
+  // Test if idx is conflict with any index in inner_idxs
+  bool ConflictInnerIndex(BlockArgument* idx);
+
+  // Number of processors
+  size_t n_cpus;
+  // The current op
+  ParallelForOp cur_op;
+  // M, N, K in inner block
+  llvm::SmallVector<BlockArgument*, numIndex> inner_idxs;
+  // M, N, K's tiles
+  size_t tiles[numIndex];
+  // Target tensors, the first two are load, the third is aggregate
+  llvm::SmallVector<Value*, numTensors> tensors;
+  // The index used by the output tensor
+  BlockArgumentSet out_idxs;
```

Comment:
@jbruestle How closely should we match LLVM style? They'd use `oldIdxs` here

Reply:
Probably we should, but if we are going to do that, we should decide where the boundary between 'style regimes' is (probably pmc/*?) and maybe see if we can get some automatic linting?  Also, we should probably make sure everyone know where the style guide is, and resolve thinks like use of references for out-vals, etc.  Seems like something to follow up on in a meeting.